bftengine/src/bcstatetransfer/SourceSelector.cpp

```diff
@@ -175,7 +182,46 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
   fetchingTimeStamp_ = 0;
   fetchRetransmissionCounter_ = 0;
   receivedValidBlockFromSrc_ = false;
+
+  LOG_INFO(logger_, "Selected new source replica " << currentReplica_);
 }
+
+void SourceSelector::updateCurrentPrimary(uint16_t newPrimary) {
+  if (currentPrimary_ == newPrimary) return;
+  ConcordAssert(isValid(newPrimary));
+  prePrepareMsgCounter_++;
+
+  auto clearCounters = [&]() {
+    nominatedPrimary_ = NO_REPLICA;
+    prePrepareMsgCounter_ = 0;
+  };
+
+  if (nominatedPrimary_ == NO_REPLICA) {
+    nominatedPrimary_ = newPrimary;
+    return;
+  }
+
+  if (prePrepareMsgCounter_ < MAX_PREPREPARE_MSGS) {
+    if (newPrimary != nominatedPrimary_) clearCounters();
+    return;
+  }
+
+  clearCounters();
+  LOG_TRACE(logger_, KVLOG(currentReplica_, currentPrimary_, newPrimary));
+
+  if (currentPrimary_ != newPrimary) {
+    if (currentPrimary_ != NO_REPLICA) {
+      addPreferredReplica(currentPrimary_);
+    }
+    // Remove immediately as retransmission timeout, rotating source may end up choosing
+    // current primary as new source
+    removePreferredReplica(newPrimary);
+  }
+
+  currentPrimary_ = newPrimary;
+  LOG_INFO(logger_, KVLOG(currentPrimary_));
```

Comment:
I don't see the reason to have 2 logs in lines 222/210.
There is no difference in terms of performance and one will be always printed in any trace level, so why not printing all the extra info? it wouldn't hurt to have extra info always, it's a rare log...
Try to merge them both to be a single log.
`LOG_INFO(logger_, "New currentPrimary selected" ..... print old currentPrimary_, new currentPrimary_ (newPrimary), and currentReplica_`